File: GitHub/expense_tracker/lib/screens/statistics/StatisticsScreen.java
import 'package:fl_chart/fl_chart.dart';
import 'package:flutter/material.dart';

import '../../models/expense_model.dart';
import '../../services/expense_service.dart';

class StatisticsScreen extends StatefulWidget {
  const StatisticsScreen({super.key});

  @override
  State<StatisticsScreen> createState() => _StatisticsScreenState();
}

class _StatisticsScreenState extends State<StatisticsScreen> {
  final ExpenseService expenseService = ExpenseService();

  List<ExpenseModel> expenses = [];
  bool loading = true;

  @override
  void initState() {
    super.initState();
    loadExpenses();
  }

  Future<void> loadExpenses() async {
    expenses = await expenseService.getExpenses();

    setState(() {
      loading = false;
    });
  }

  double get totalExpense {
    double total = 0;

    for (var e in expenses) {
      total += e.amount;
    }

    return total;
  }

  double categoryTotal(String category) {
    double total = 0;

    for (var e in expenses) {
      if (e.category == category) {
        total += e.amount;
      }
    }

    return total;
  }

  @override
  Widget build(BuildContext context) {
    if (loading) {
      return const Scaffold(
        body: Center(
          child: CircularProgressIndicator(),
        ),
      );
    }

    return Scaffold(
      appBar: AppBar(
        backgroundColor: Colors.red,
        title: const Text(
          "Statistics",
          style: TextStyle(color: Colors.white),
        ),
      ),
      body: SingleChildScrollView(
        padding: const EdgeInsets.all(20),
        child: Column(
          children: [
            Card(
              elevation: 4,
              child: Padding(
                padding: const EdgeInsets.all(20),
                child: Column(
                  children: [
                    const Text(
                      "Total Expense",
                      style: TextStyle(fontSize: 18),
                    ),
                    const SizedBox(height: 10),
                    Text(
                      "Rs ${totalExpense.toStringAsFixed(0)}",
                      style: const TextStyle(
                        fontSize: 32,
                        fontWeight: FontWeight.bold,
                        color: Colors.red,
                      ),
                    ),
                  ],
                ),
              ),
            ),

            const SizedBox(height: 30),

            SizedBox(
              height: 260,
              child: PieChart(
                PieChartData(
                  centerSpaceRadius: 45,
                  sectionsSpace: 3,
                  sections: [
                    PieChartSectionData(
                      value: categoryTotal("Food"),
                      color: Colors.red,
                      title: "Food",
                    ),
                    PieChartSectionData(
                      value: categoryTotal("Travel"),
                      color: Colors.orange,
                      title: "Travel",
                    ),
                    PieChartSectionData(
                      value: categoryTotal("Shopping"),
                      color: Colors.blue,
                      title: "Shop",
                    ),
                    PieChartSectionData(
                      value: categoryTotal("Bills"),
                      color: Colors.green,
                      title: "Bills",
                    ),
                  ],
                ),
              ),
            ),

            const SizedBox(height: 25),

            buildCard(
              Icons.fastfood,
              "Food",
              Colors.red,
              categoryTotal("Food"),
            ),

            buildCard(
              Icons.directions_car,
              "Travel",
              Colors.orange,
              categoryTotal("Travel"),
            ),

            buildCard(
              Icons.shopping_cart,
              "Shopping",
              Colors.blue,
              categoryTotal("Shopping"),
            ),

            buildCard(
              Icons.receipt_long,
              "Bills",
              Colors.green,
              categoryTotal("Bills"),
            ),
          ],
        ),
      ),
    );
  }

  Widget buildCard(
      IconData icon,
      String title,
      Color color,
      double amount,
      ) {
    return Card(
      child: ListTile(
        leading: CircleAvatar(
          backgroundColor: color,
          child: Icon(
            icon,
            color: Colors.white,
          ),
        ),
        title: Text(title),
        trailing: Text(
          "Rs ${amount.toStringAsFixed(0)}",
          style: const TextStyle(
            fontWeight: FontWeight.bold,
            fontSize: 17,
          ),
        ),
      ),
    );
  }
}

File: GitHub/expense_tracker/lib/screens/statistics/statistics_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:fl_chart/fl_chart.dart';
import 'package:expense_tracker/providers/expense_provider.dart';
import 'package:expense_tracker/utils/app_colors.dart';

class StatisticsScreen extends StatelessWidget {
  const StatisticsScreen({super.key});

  @override
  Widget build(BuildContext context) {
    final provider = Provider.of<ExpenseProvider>(context);
    final isDark = provider.isDarkMode;
    final _ = provider.activeCurrencySymbol;
    final rate = provider.currencyMultiplier;

    final totalIncome = provider.totalIncome * rate;
    final totalExpense = provider.totalExpense * rate;
    final _ = totalIncome + totalExpense;

    return Scaffold(
      backgroundColor: AppColors.getBackground(isDark),
      appBar: AppBar(
        backgroundColor: AppColors.getBackground(isDark),
        elevation: 0,
        title: Text("Analytics", style: TextStyle(color: AppColors.getTextDark(isDark), fontWeight: FontWeight.bold)),
        centerTitle: true,
      ),
      body: totalIncome == 0 && totalExpense == 0
          ? Center(child: Text("No data rows yet.", style: TextStyle(color: AppColors.getTextLight(isDark))))
          : SingleChildScrollView(
        padding: const EdgeInsets.all(24.0),
        child: Column(
          children: [
            SizedBox(
              height: 200,
              child: PieChart(
                PieChartData(
                  sectionsSpace: 4,
                  centerSpaceRadius: 55,
                  sections: [
                    PieChartSectionData(color: AppColors.greenSmooth, value: totalIncome > 0 ? totalIncome : 1, title: 'Income', radius: 50, titleStyle: const TextStyle(color: Colors.white)),
                    PieChartSectionData(color: AppColors.primary, value: totalExpense > 0 ? totalExpense : 1, title: 'Expense', radius: 50, titleStyle: const TextStyle(color: Colors.white)),
                  ],
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
